### ISSUE TITLE
tkz-euclide/pgfplots macros/commands are not recognized

### ISSUE THREAD

**Issue by LoganDark** [texstudio-org/texstudio#752]
### Environment
- TeXstudio: 2.12.16
- Qt: 5.12.3 R
- OS: macOS 10.14.5
- TeX distribution: `pdfTeX, Version 3.14159265-2.6-1.40.17 (TeX Live 2016)`

### Expected behavior
Things like `tkz-euclide` macros/commands are recognized

### Actual behavior
![image](https://user-images.githubusercontent.com/4723091/66791199-09551c80-eea8-11e9-8dda-ac2eb1aa9bc8.png)

This happens with tkz-euclide, pgfplots, and some other libraries

### How to reproduce
```TeX
\documentclass[12pt]{article}
\usepackage[utf8]{inputenc}
\usepackage{helvet,parskip,amsmath}
\usepackage[a4paper, margin=1in]{geometry}

\title{Minimal Working Example}
\author{LoganDark}
\date{October 2019}

\renewcommand{\familydefault}{\sfdefault}
\renewcommand{\baselinestretch}{1.25}

\usepackage{tikz,tkz-euclide}
\usetkzobj{all}

\begin{document}
	\begin{tikzpicture}[scale=1]
		\tkzInit[xmax=7,ymax=7]
		%\tkzAxeXY
		%\tkzGrid
		
		\begin{pgfinterruptboundingbox}
			\tkzDefPoints{3/3/C,6/3/R,4/6/sA,6/2/sB}
			\tkzInterLC(sA,sB)(C,R)
			\tkzGetPoints{A}{B}
			\tkzDefPointBy[homothety=center B ratio 1.25](A)
			\tkzGetPoint{Ae}
			\tkzDefPointBy[homothety=center A ratio 1.25](B)
			\tkzGetPoint{Be}
		\end{pgfinterruptboundingbox}
		
		\tkzDrawSegment(A,B)
		\tkzDrawSegments[style=dashed](Ae,A B,Be)
		\tkzDrawCircle(C,R)
		\tkzDrawPoints(A,B)
		\tkzLabelPoints[right](B)
		\tkzLabelPoints[above right](A)
	\end{tikzpicture}
\end{document}
```
![image](https://user-images.githubusercontent.com/4723091/66791490-81701200-eea9-11e9-8f7a-1aab187fdb3d.png)

**Comment by dca00** [comment]
Texstudio probably should not strive to recognize all macros of all packages.

**Comment by LoganDark** [comment]
Then why does it have to mark them wrong?

**Comment by muzimuzhi** [comment]
They are just marked as "unrecognized command/environment" (when cursor is hover on the commands/environment).

Texstudio uses cwl files to highlight the predefined commands from loaded packages. It is welcome to to contribute to texstudio project by "adding missing cwl files". The related document is [Sec. 4.13 of user manual](http://texstudio.sourceforge.net/manual/current/usermanual_en.html#CWLDESCRIPTION).

**Comment by sunderme** [comment]
usually txs can parse unknown styles automatically to generate a provisional cwl in .config/texstudio/completion/autogenerated
This fails occasionally as .sty files are notoriously hard to parse.
You can generate a user file (see manual) which lists the commands, possibly just by extending whatever cwl txs has managed to generate.

**Comment by timhoffm** [comment]
See also [FAQ: Why does TeXstudio mark a command as missing](https://github.com/texstudio-org/texstudio/wiki/Frequently-Asked-Questions#why-does-texstudio-mark-a-command-as-unrecognized) and the links therein.

You are welcome to contribute a `.cwl` with the command definitions of `tkz-euclide`.

**Comment by LoganDark** [comment]
@timhoffm this is still an issue until the cwl files are added. I'm not sure if I can add them myself but I think closing the issue is a bit drastic. This feels kinda like a "we don't care and you can do it yourself" so far

**Comment by timhoffm** [comment]
@LoganDark I'm sorry if this feels a bit drastic. We do have a .sty file parser that can automatically create basic cwl files, but it is not possible to identify all ways a command can be defined in .sty files. Additionally that parser cannot attach semantic information to the commands.

There are edge cases which require manual work. Given the limited bandwidth of core developers, we cannot promise to support these as a core feature of TeXstudio. We have to rely on community contributed .cwl files here, and we welcome these very much. But on the other hand, I don't see it as a bug of TeXstudio if they are not available. This part of language detection is inherently incomplete.

**Comment by dbitouze** [comment]
>But on the other hand, I don't see it as a bug of TeXstudio if they are not available

Indeed, but could be considered as "enhancement".

**Comment by timhoffm** [comment]
I've introduced a new label "cwl" and reopened.

This will not be targeted by core developers, but everybody from the community is very welcome to contribute a hand-crafted cwl file.

**Comment by LoganDark** [comment]
> Indeed, but could be considered as "enhancement".

At this point it would be better to create a new issue about the missing cwl files. From what I can see the devs don't feel like it would be worth the effort for TeXstudio to automatically look for commands, so an issue about the editor not discovering commands is basically a 'works as intendeed wontfix'.

**Comment by timhoffm** [comment]
> From what I can see the devs don't feel like it would be worth the effort for TeXstudio to automatically look for commands, so an issue about the editor not discovering commands is basically a 'works as intendeed wontfix'.

I'm sorry, but that's absolutely not true. We do have put quite some effort in that automatism, but TeX is too complex to make it work for all cases.

**Comment by LoganDark** [comment]
> TeX is too complex to make it work for all cases.

Would you be willing to look into the specific method tkz-euclide uses to define its commands then? I understand it might not be possible to solve the issue for every package but the packages included in the distribution should work and be recognized, right?

**Comment by timhoffm** [comment]
> Would you be willing to look into the specific method tkz-euclide uses to define its commands then?

I'm sorry, but no. Further improving the .sty parser this is a rabbit hole I won't dive into. I have too little time for this.

>  I understand it might not be possible to solve the issue for every package but the packages included in the distribution should work and be recognized, right?

Even that is a goal too high, unfortunately.

**Comment by mbertucci47** [comment]
There are now cwl files for each of the `tkz-*` packages in the completion folder. I imagine this issue should be closed now.

**Comment by LoganDark** [comment]
@mbertucci47 If the tkz-euclide commands provided in the example code are recognized, then the issue can be closed.

**Comment by mbertucci47** [comment]
With the small fix in pull request #1674, the commands in the example are all recognized. However, the various `pgf` completion files are still very much incomplete.

**Comment by LoganDark** [comment]
> With the small fix in pull request #1674, the commands in the example are all recognized. However, the various `pgf` completion files are still very much incomplete.

Feel free to include the magic words `closes #752` in your issue text to link that PR to this issue
